### ISSUE TITLE
startSpan needs to return Context instead of Span

### ISSUE THREAD

**Issue by anuraaga** [open-telemetry/opentelemetry-java-instrumentation#1481]
@marcingrzejszczak is probably one of our first serious users of baggage :) I was helping him debug an issue and noticed that currently it seems impossible to use baggage with our instrumentation because `startSpan` extracts from the request, but only returns a `Span`. The baggage information that was extracted is therefore lost (except for as a private field of `Span`, it's parent).

We need `startSpan` to always return a `Context` to make sure that span is together with its baggage. I think this means that we will remove all usage of `currentContextWith` type of pattern in our repo, since it's really the job of the tracer's to produce the context, not to use the current context directly IIUC. Calling `Span.fromContext` everywhere will be tedious, but it's what the spec wants :)

**Comment by bogdandrutu** [comment]
> @marcingrzejszczak is probably one of our first serious users of baggage :) I was helping him debug an issue and noticed that currently it seems impossible to use baggage with our instrumentation because startSpan extracts from the request, but only returns a Span. The baggage information that was extracted is therefore lost (except for as a private field of Span, it's parent).

Can this be clarified, what exactly needs to be changed in `baggage`.

**Comment by bogdandrutu** [comment]
Also please provide a code example, because this is not clear (at least to me).

**Comment by anuraaga** [comment]
> Can this be clarified, what exactly needs to be changed in baggage.

I don't think anything needs to change with baggage, just instrumentation. Our instrumentation for any server framework, e.g. spring or gRPC, extract from headers here

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpServerTracer.java#L59

But the result of the extraction isn't visible to the caller of `startSpan` - instead when we mount a context, we use current context, which doesn't have baggage

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpServerTracer.java#L80

What we need to do is mount `extracted.with(span)` though and to do that, without running extraction logic twice at least, is to return `Context` from `startSpan` I think.

**Comment by bogdandrutu** [comment]
This is related to https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1429.

**Comment by iNikem** [comment]
@bogdandrutu I don't see how moving from "tracers" to "instrumenters" is related to this, very localised, issue of one method's return type?

**Comment by marcingrzejszczak** [comment]
> Also please provide a code example, because this is not clear (at least to me).

`io.opentelemetry.instrumentation.api.tracer.HttpServerTracer#startSpan(...)`
```java
public Span startSpan(REQUEST request, CONNECTION connection, String spanName) {
    return startSpan(request, connection, spanName, -1);
  }

  public Span startSpan(
      REQUEST request, CONNECTION connection, String spanName, long startTimestamp) {
    // extract(...) retrieves context with baggage but that context is not propagated
    Span.Builder builder =
        tracer.spanBuilder(spanName).setSpanKind(SERVER).setParent(extract(request, getGetter()));

    if (startTimestamp >= 0) {
      builder.setStartTimestamp(startTimestamp);
    }

    Span span = builder.startSpan();
    onConnection(span, connection);
    onRequest(span, request);
    onConnectionAndRequest(span, connection, request);
    // this span doesn't have baggage propagated
    return span;
  }
```

how it could be fixed (this is just to show the problem - ultimately it should be done as @anuraaga suggests)

```java
public Span startSpan(REQUEST request, CONNECTION connection, String spanName) {
    return startSpan(request, connection, spanName, -1);
  }

  public Span startSpan(
      REQUEST request, CONNECTION connection, String spanName, long startTimestamp) {
    Context ctx = extract(request, getGetter());
    Span.Builder builder =
        tracer.spanBuilder(spanName).setSpanKind(SERVER).setParent(ctx);

    if (startTimestamp >= 0) {
      builder.setStartTimestamp(startTimestamp);
    }

    Span span = builder.startSpan();
    onConnection(span, connection);
    onRequest(span, request);
    onConnectionAndRequest(span, connection, request);
    // this span does have baggage propagated
    return ctx.with(span);
  }
```

**Comment by iNikem** [comment]
Closed by #1525
